File: py/tests/test_negative_zero.py
"""No public estimator ever reports a negative zero.

A sample holding both ``+0.0`` and ``-0.0`` used to make the reported value depend on which
of the two the sort happened to leave in the selected position: comparison cannot separate
them, so the position was settled by the sorting algorithm rather than by the data, and the
seven ports each bring their own sort. ``center([0.0, -0.0, 0.0, -0.0, 1.0])`` came out of
this port as ``-0.0`` and out of Go as ``+0.0``, against an ``exact`` conformance class that
promises identical bits from identical inputs.

Every estimator now sheds the sign on the way out, so these tests compare PAYLOADS: ``-0.0
== 0.0`` is true, and an equality assertion would pass on exactly the results this suite
exists to reject.

The samples below are the ones that reached a negative zero before the fix, plus a sweep
that holds the invariant for the estimators where it is currently unreachable. Those are
unreachable by proof, not by construction (``spread`` is an absolute difference, ``ratio``
is an exponential), and a proof nobody re-checks is how the next divergence gets in.
"""

import math

import pytest
from binary64 import assert_identical, fmt, payload

from pragmastat import (
    Sample,
    center,
    center_bounds,
    disparity,
    disparity_bounds,
    ratio,
    ratio_bounds,
    shift,
    shift_bounds,
    spread,
    spread_bounds,
)
from pragmastat.estimators import _avg_spread, _avg_spread_bounds, _new_bounds, _normalize_zero
from pragmastat.measurement_unit import NUMBER_UNIT

NEGATIVE_ZERO = 0x8000000000000000

# Every one of these estimates to zero, and every one of them selected a -0.0 before the fix.
MIXED_ZERO_SAMPLES = [
    [0.0, -0.0, 0.0, -0.0, 1.0],
    [-0.0, -0.0],
    [-0.0, 0.0],
    [0.0, -0.0],
    [-0.0, -0.0, -0.0],
    [-1.0, 1.0],
    [-2.0, -0.0, 2.0],
]

MIXED = [0.0, -0.0, 0.0, -0.0, 1.0, -1.0]
POSITIVE = [1.0, 2.0, 3.0, 4.0, 5.0, 6.0]


def assert_positive_zero(value, what):
    assert_identical(value, 0.0, what)


@pytest.mark.parametrize("x", MIXED_ZERO_SAMPLES)
def test_center_reports_a_positive_zero(x):
    assert_positive_zero(center(x), f"center({x})")


@pytest.mark.parametrize("x", MIXED_ZERO_SAMPLES)
def test_center_reports_a_positive_zero_for_a_sample(x):
    assert_positive_zero(center(Sample(x)).value, f"center(Sample({x}))")


def test_center_bounds_reports_positive_zeros():
    bounds = center_bounds(MIXED, 0.3)
    assert_positive_zero(bounds.lower, "center_bounds lower")
    assert_positive_zero(bounds.upper, "center_bounds upper")


def test_center_bounds_reports_positive_zeros_for_a_sample():
    bounds = center_bounds(Sample(MIXED), 0.3)
    assert_positive_zero(bounds.lower, "center_bounds lower")
    assert_positive_zero(bounds.upper, "center_bounds upper")


@pytest.mark.parametrize(
    ("x", "y"),
    [
        ([-0.0], [0.0]),
        ([-0.0, 1.0], [0.0, 1.0]),
        ([-0.0, -0.0], [0.0, 0.0]),
    ],
)
def test_shift_reports_a_positive_zero(x, y):
    assert_positive_zero(shift(x, y), f"shift({x}, {y})")


def test_shift_bounds_reports_positive_zeros():
    # A single pair takes the degenerate x[0] - y[0] path, where -0.0 - +0.0 is -0.0.
    bounds = shift_bounds([-0.0], [0.0], 1.0)
    assert_positive_zero(bounds.lower, "shift_bounds lower")
    assert_positive_zero(bounds.upper, "shift_bounds upper")


def test_disparity_reports_a_positive_zero():
    # Shift selects the -0.0 difference; the average spread is positive, so the sign survived.
    assert_positive_zero(disparity([-0.0, 1.0], [0.0, 1.0]), "disparity")


def public_outputs():
    """Every public estimator, evaluated on input that carries a signed zero."""
    yield "center", center(MIXED)
    yield "spread", spread(MIXED)
    yield "shift", shift(MIXED, MIXED)
    yield "ratio", ratio(POSITIVE, POSITIVE)
    yield "disparity", disparity(MIXED, MIXED)
    yield "avg_spread", _avg_spread(Sample(MIXED), Sample(MIXED)).value

    named_bounds = [
        ("center_bounds", center_bounds(MIXED, 0.3)),
        ("spread_bounds", spread_bounds(MIXED, 0.5, "seed")),
        ("shift_bounds", shift_bounds(MIXED, MIXED, 0.5)),
        ("ratio_bounds", ratio_bounds(POSITIVE, POSITIVE, 0.5)),
        ("disparity_bounds", disparity_bounds(MIXED, MIXED, 0.9, "seed")),
        ("avg_spread_bounds", _avg_spread_bounds(Sample(MIXED), Sample(MIXED), 0.9, "seed")),
    ]
    for name, bounds in named_bounds:
        yield f"{name} lower", bounds.lower
        yield f"{name} upper", bounds.upper


@pytest.mark.parametrize(("name", "value"), list(public_outputs()))
def test_no_estimator_reports_a_negative_zero(name, value):
    assert payload(value) != NEGATIVE_ZERO, f"{name} reported a negative zero: {fmt(value)}"


def test_inputs_keep_their_negative_zeros():
    # Only outputs are normalized: a sample must still be able to CONTAIN a -0.0.
    x = [0.0, -0.0, 0.0, -0.0, 1.0]
    center(x)
    assert payload(x[1]) == NEGATIVE_ZERO, f"input was rewritten: {fmt(x[1])}"

    sample = Sample(x)
    center(sample)
    assert payload(sample.values[1]) == NEGATIVE_ZERO, f"sample values were rewritten: {fmt(sample.values[1])}"
    assert NEGATIVE_ZERO in [payload(v) for v in sample.sorted_values], "sorted values were rewritten"


@pytest.mark.parametrize(
    "value",
    [1.0, -1.0, 1e-320, -1e-320, math.inf, -math.inf, math.nan, 5e-324, -5e-324],
)
def test_normalization_leaves_every_non_zero_alone(value):
    assert_identical(_normalize_zero(value), value, f"_normalize_zero({value!r})")


@pytest.mark.parametrize("value", [0.0, -0.0])
def test_normalization_maps_both_zeros_to_the_positive_one(value):
    assert_identical(_normalize_zero(value), 0.0, f"_normalize_zero({value!r})")


def test_new_bounds_normalizes_both_endpoints():
    bounds = _new_bounds(-0.0, -0.0, NUMBER_UNIT)
    assert_positive_zero(bounds.lower, "lower")
    assert_positive_zero(bounds.upper, "upper")


def test_new_bounds_leaves_infinite_endpoints_alone():
    bounds = _new_bounds(-math.inf, math.inf, NUMBER_UNIT)
    assert_identical(bounds.lower, -math.inf, "lower")
    assert_identical(bounds.upper, math.inf, "upper")

File: r/pragmastat/tests/testthat/test-negative-zero.R
# Regression: on a sample carrying both +0 and -0, nothing in the data decided
# which of them landed in the position an estimator selects. Comparison cannot
# separate the two, so the sort picked, every port sorts its own way, and the
# ports returned different bits from the same input. Every estimator now
# normalizes the sign away on the way out (see normalize_zero).
#
# -0 == 0 is TRUE, so an equality check proves nothing here. These tests compare
# the 64-bit payload through double_bits() from helper-reference-tests.R.

POSITIVE_ZERO <- "0x0000000000000000"
NEGATIVE_ZERO <- "0x8000000000000000"

test_that("the payload check separates the two zeros", {
  expect_identical(double_bits(0.0), POSITIVE_ZERO)
  expect_identical(double_bits(-0.0), NEGATIVE_ZERO)
})

test_that("center returns positive zero on samples carrying signed zeros", {
  samples <- list(
    c(0.0, -0.0, 0.0, -0.0, 1.0),
    c(-0.0, -0.0),
    c(-0.0, 0.0),
    c(0.0, -0.0),
    c(-0.0, -0.0, -0.0),
    c(-1.0, 1.0),
    c(-2.0, -0.0, 2.0)
  )
  for (x in samples) {
    label <- paste0("c(", paste(sprintf("%.17g", x), collapse = ", "), ")")
    expect_identical(double_bits(center(x)), POSITIVE_ZERO,
      info = paste("center", label)
    )
    expect_identical(double_bits(unwrap_value(center(Sample$new(x)))), POSITIVE_ZERO,
      info = paste("center Sample", label)
    )
  }
})

test_that("center_bounds returns positive zero on both endpoints", {
  x <- c(0.0, -0.0, 0.0, -0.0, 1.0, -1.0)
  raw_bounds <- center_bounds(x, 0.3)
  expect_identical(double_bits(raw_bounds$lower), POSITIVE_ZERO)
  expect_identical(double_bits(raw_bounds$upper), POSITIVE_ZERO)

  sample_bounds <- center_bounds(Sample$new(x), 0.3)
  expect_identical(double_bits(sample_bounds$lower), POSITIVE_ZERO)
  expect_identical(double_bits(sample_bounds$upper), POSITIVE_ZERO)
})

# The samples below each produced a negative zero before the fix, one per
# estimator that can reach a zero at all. spread, avg_spread and their bounds
# are absent because they cannot: spread > 0 is an enforced assumption, and the
# spread bounds are order statistics of absolute differences, which are never
# negative zeros.
test_that("shift and disparity return positive zero", {
  x <- c(1.0, -1.0, -0.0, -0.0, -1.0, 0.0)
  y <- c(-2.0, -0.0, 0.0)
  expect_identical(double_bits(shift(x, y)), POSITIVE_ZERO)
  expect_identical(double_bits(disparity(x, y)), POSITIVE_ZERO)
})

test_that("shift_bounds returns positive zero endpoints", {
  # The single-pair branch: one x against one y, the difference returned as both
  # endpoints without going through the quantile search.
  degenerate <- shift_bounds(c(-0.0), c(0.0), 1.0)
  expect_identical(double_bits(degenerate$lower), POSITIVE_ZERO)
  expect_identical(double_bits(degenerate$upper), POSITIVE_ZERO)

  searched <- shift_bounds(c(-0.0), c(0.0, 2.0, 1.0, -2.0), 0.5)
  expect_identical(double_bits(searched$upper), POSITIVE_ZERO)
})

test_that("center_bounds and disparity_bounds return positive zero endpoints", {
  cb <- center_bounds(c(-0.0, -1.0), 0.5)
  expect_identical(double_bits(cb$upper), POSITIVE_ZERO)

  db <- disparity_bounds(
    c(-0.0, 0.0, 2.0, -2.0, -0.0, 1.0),
    c(0.0, 2.0, 2.0, 2.0, 1.0, 0.0),
    0.9,
    seed = "negative-zero"
  )
  expect_identical(double_bits(db$upper), POSITIVE_ZERO)
})

test_that("negative zeros stay legal in the input", {
  # Only outputs are normalized: a sample may still contain -0, and it still
  # compares equal to +0, so the estimate is unaffected.
  expect_identical(double_bits(center(c(-0.0, 4.0))), double_bits(center(c(0.0, 4.0))))
  expect_identical(double_bits(shift(c(-0.0, 4.0), c(1.0))), double_bits(shift(c(0.0, 4.0), c(1.0))))
})
